### PR TITLE
[C++ API] expose C++ task failure event

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?
- Bugfix: return `Status::CreationTaskError` when actor creation task failed
- Bugfix: Rename event directory from "event" to "events" which is the matching name defined in event agent. https://github.com/ray-project/ray/blob/master/dashboard/modules/event/event_agent.py#L25
- Add `RAY_CPP_TASK_FAILED` event to expose cpp task failure.

### CODE REVIEW

**Review by qicosmos — APPROVED** [review]
LGTM

File: cpp/src/ray/runtime/task/task_executor.cc [hunk 1]
```diff
@@ -159,6 +161,11 @@ Status TaskExecutor::ExecuteTask(
   if (!status.ok()) {
     if (status.IsIntentionalSystemExit()) {
       return status;
+    } else {
+      RAY_EVENT(ERROR, EL_RAY_CPP_TASK_FAILED)
+              .WithField("task_type", TaskType_Name(task_type))
+              .WithField("function_name", func_name)
+          << "C++ task failed: " << status.ToString();
```
Comment (kfstorm) [inline cpp/src/ray/runtime/task/task_executor.cc:168]:
```suggestion
          << "Failed to execute the C++ task " << ray_function.GetFunctionDescriptor()->ToString() << ": " << status.ToString();
```

**Review by kfstorm — APPROVED** [review]
LGTM.

File: cpp/src/ray/test/cluster/counter.cc [hunk 2]
```diff
@@ -28,6 +28,10 @@ Counter::Counter(int init) {
 
 Counter *Counter::FactoryCreate() { return new Counter(0); }
 
+Counter *Counter::FactoryCreateException() {
+  throw std::invalid_argument("creation error");
```
Comment (raulchen) [inline cpp/src/ray/test/cluster/counter.cc:32]: better throw the exception in the constructor

File: cpp/src/ray/runtime/task/task_executor.cc (hunk 1 above)
Comment (SongGuyang) [reply cpp/src/ray/runtime/task/task_executor.cc:168]: We already set func_name the custom field, so here we don't need to print duplicated information to message. And as the unique field,  func_name could represent `FunctionDescriptor`.

**Pull request merged**